### PR TITLE
common/config: eliminate config_t::set_val unsafe option

### PR DESCRIPTION
**Pull request by Liuchang0812**
Fixes: http://tracker.ceph.com/issues/19106

Signed-off-by: liuchang0812 <liuchang0812@gmail.com>

### CODE REVIEW

**Comment by Liuchang0812** [comment]
jenkins test this please

**Comment by Liuchang0812** [comment]
Working fail test cases

**Comment by Liuchang0812** [comment]
:rofl: I need help. It seems that `ctest` could not find some *.so files, but it is ok when I try to run unittest binary directly.  What should I do to fix this problem? @tchaikov 

more logs as following:
```
➜  build git:(wip-19106-eliminate-unsafe-option) ✗ ctest -VV -R unittest_compression
UpdateCTestConfiguration  from :/home/liuchang/WorkSpace/ceph/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/liuchang/WorkSpace/ceph/build/DartConfiguration.tcl
Test project /home/liuchang/WorkSpace/ceph/build
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 85
    Start 85: unittest_compression

85: Test command: /home/liuchang/WorkSpace/ceph/build/bin/unittest_compression
85: Environment variables:
85:  CEPH_ROOT=/home/liuchang/WorkSpace/ceph
85:  CEPH_BIN=/home/liuchang/WorkSpace/ceph/build/bin
85:  CEPH_LIB=/home/liuchang/WorkSpace/ceph/build/lib
85:  CEPH_BUILD_DIR=/home/liuchang/WorkSpace/ceph/build
85:  LD_LIBRARY_PATH=/home/liuchang/WorkSpace/ceph/build/lib
85:  PATH=/home/liuchang/WorkSpace/ceph/build/bin:/home/liuchang/WorkSpace/ceph/src:/home/liuchang/bin:/home/liuchang/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
85:  PYTHONPATH=/home/liuchang/WorkSpace/ceph/build/lib/cython_modules/lib.2:/home/liuchang/WorkSpace/ceph/src/pybind
85:  CEPH_BUILD_VIRTUALENV=/tmp
85: Test timeout computed to be: 3600
85: 2017-03-01 14:27:35.522809 7fc7e144af00 -1 did not load config file, using default settings.
85: [==========] Running 68 tests from 3 test cases.
85: [----------] Global test environment set-up.
85: [----------] 3 tests from ZlibCompressor
85: [ RUN      ] ZlibCompressor.zlib_isal_compatibility
85: *** Caught signal (Segmentation fault) **
85:  in thread 7fc7e144af00 thread_name:unittest_compre
85: 2017-03-01 14:27:35.525576 7fc7e144af00 -1 load failed dlopen(): "/usr/local/lib/ceph/compressor/libceph_zlib.so: cannot open shared object file: No such file or directory" or "/usr/local/lib/ceph/libceph_zlib.so: cannot open shared object file: No such file or directory"
85: 2017-03-01 14:27:35.525591 7fc7e144af00 -1 create cannot load compressor of type zlib


➜  build git:(wip-19106-eliminate-unsafe-option) ✗ ./bin/unittest_compression
2017-03-01 14:29:10.801193 7fca8f4b2f00 -1 WARNING: all dangerous and experimental features are enabled.
2017-03-01 14:29:10.801358 7fca8f4b2f00 -1 WARNING: all dangerous and experimental features are enabled.
2017-03-01 14:29:10.802573 7fca8f4b2f00 -1 WARNING: all dangerous and experimental features are enabled.
[==========] Running 68 tests from 3 test cases.
[----------] Global test environment set-up.
[----------] 3 tests from ZlibCompressor
[ RUN      ] ZlibCompressor.zlib_isal_compatibility
[       OK ] ZlibCompressor.zlib_isal_compatibility (0 ms)
[ RUN      ] ZlibCompressor.isal_compress_zlib_decompress_random
[       OK ] ZlibCompressor.isal_compress_zlib_decompress_random (556 ms)
[ RUN      ] ZlibCompressor.isal_compress_zlib_decompress_walk
[       OK ] ZlibCompressor.isal_compress_zlib_decompress_walk (451 ms)

```

**Comment by Liuchang0812** [comment]
@tchaikov Do you have time to take a look ?

**Comment by Liuchang0812** [comment]
ok, I found that `PluginRegistry::load` depends on `cct->_conf->plugin_dir`.

**Comment by gregsfortytwo** [comment]
It looks like this just eliminates the safe/unsafe tracking. I think the point of that card is to make all config sets have safe behavior?

**Comment by Liuchang0812** [comment]
@gregsfortytwo thanks for your comment. 

I think that i don't eliminates the safe/unsafe tracking.  `md_config_t` has `is_safe()` [1] function which makes a type check. integer and floating point options considered thread safe by default. we could use `SAFE_OPTION` to declare a safe config as [2] .  `OPTION` with string type is unsafe default. So we do not need the `safe` parameter in `md_config_t::set_val` [3] .

I am not sure whether i get correct point of the trello card.  @liewegas Would you mind taking a look at this question?

[1]:https://github.com/ceph/ceph/blob/8d179d2a14360b6875a7a09215f3ea07768cf568/src/common/config.cc#L804
[2]:https://github.com/ceph/ceph/blob/8d179d2a14360b6875a7a09215f3ea07768cf568/src/common/config_opts.h#L1365
[3]:https://github.com/ceph/ceph/blob/4fe7e2a458a1521839bc390c2e3233dd809ec3ac/src/common/config.h#L210

**Comment by liewegas** [comment]
The point of hte card is to fix all of the callers that pass safe=false to set_val.  They currently have to do that because the system is past the init phsae when we're free to update these at will and the options don't have observers.  Each caller needs to be looked at individually to figure out if/how that's possible.  Only after all the callers are fixed can we drop the option!

For example, in global_init.cc, instead of changing the setuser value and then checking it 15 lines down, use getuid() to decide whether whether to check the fields at all.

**Comment by Liuchang0812** [comment]
@liewegas :+1:  thanks your explaining. I get its point now. 

in global_init.cc, We could configure `conf->setuid` forced because of that `getuid` function guards the function atomic, other threads can't visit those values. is it right?

**Comment by Liuchang0812** [comment]
In other words, We could remove safe param if we have used SAFE_OPTION to declare all options that modified in this patch.

**Comment by liewegas** [comment]
We could, but that doesn't improve the situation.  The problem is that we occasionally see valgrind errors showing a corrupted std::string in the config structure.  The point of removing the 'unsafe' set_val is to structurally make it impossible for the config to be modified once it is loaded without using the observer framework (which is allows config options to be read under a lock).

The first step is to fix up any users that simply don't need to use it.  The setuid one is an easy example.. there's no reason to change the string at all; we can just restructure the calling code.

File: src/test/msgr/test_async_networkstack.cc [hunk 1]
```diff
@@ -41,17 +41,17 @@ class NetworkWorkerTest : public ::testing::TestWithParam<const char*> {
...
     cerr << __func__ << " start set up " << GetParam() << std::endl;
     if (strncmp(GetParam(), "dpdk", 4)) {
-      g_ceph_context->_conf->set_val("ms_type", "async+posix", false, false);
+      g_ceph_context->_conf->set_val("ms_type", "async+posix", false);
       addr = "127.0.0.1:15000";
       port_addr = "127.0.0.1:15001";
     } else {
-      g_ceph_context->_conf->set_val("ms_type", "async+dpdk", false, false);
-      g_ceph_context->_conf->set_val("ms_dpdk_debug_allow_loopback", "true", false, false);
-      g_ceph_context->_conf->set_val("ms_async_op_threads", "2", false, false);
```
Comment (Liuchang0812) [inline src/test/msgr/test_async_networkstack.cc]: ms_async_op_threads is a int config. it's safe. no need to pass `safe=false`

File: src/test/msgr/test_async_networkstack.cc [hunk 2]
```diff
@@ -41,17 +41,17 @@ class NetworkWorkerTest : public ::testing::TestWithParam<const char*> {
   void SetUp() override {
     cerr << __func__ << " start set up " << GetParam() << std::endl;
     if (strncmp(GetParam(), "dpdk", 4)) {
-      g_ceph_context->_conf->set_val("ms_type", "async+posix", false, false);
+      g_ceph_context->_conf->set_val("ms_type", "async+posix", false);
       addr = "127.0.0.1:15000";
       port_addr = "127.0.0.1:15001";
     } else {
-      g_ceph_context->_conf->set_val("ms_type", "async+dpdk", false, false);
-      g_ceph_context->_conf->set_val("ms_dpdk_debug_allow_loopback", "true", false, false);
```
Comment (Liuchang0812) [inline src/test/msgr/test_async_networkstack.cc]: OPTION(ms_dpdk_debug_allow_loopback, OPT_BOOL, false)

It's OPT_BOOL, no need to pass `safe=false`

File: src/msg/simple/Pipe.cc [hunk 3]
```diff
@@ -209,11 +209,13 @@ void Pipe::start_reader()
 
 void Pipe::maybe_start_delay_thread()
 {
-  if (!delay_thread &&
```
Comment (Liuchang0812) [inline src/msg/simple/Pipe.cc]: this line is too long. split to 3 lines.

**Comment by Liuchang0812** [comment]
@liewegas 

1. `setuser/setgroup`: they are no need to modify. We could just use `getuid` to decide whether to do the work.
2. `ms_async_op_threads/ms_dpdk_debug_allow_loopback`: they are POD types. We could eliminate `safe=false` option in set_val.
3. other options are modified in run-time. They are converted to SAFE_OPTION. 

Now, there are no functions that need `safe=false` option. 

Please review this commit carefully: https://github.com/ceph/ceph/pull/13687/commits/bb72c7600ab8cc8d9a8c3e66671a250dd6c3a241


There are some works we could do in future:

1.  Give template function `template <typename T> T md_config_t::get_val`  default type `std::string`. Now there are almost `g_conf->get_val<std::string>("")` code. We could make code more readable.
2.  For OPTION_STR, We should give explicit tips if user try to `set_val` them, rather then that do not apply changes now. It would be better if we could check it in compile time(use static_assert?).
3. We should disable all `g_conf->foo` calling mode of OPTION_STR. We couldn't do it now, because there are too much code need to be changed. Maybe we could introduce macro `FLAGS_foo` like `gflags`?

File: src/global/global_init.cc [hunk 4]
```diff
@@ -166,106 +166,107 @@ global_init(std::vector < const char * > *alt_def_args,
...
-	  exit(1);
-	}
-	uid = p->pw_uid;
-	gid = p->pw_gid;
-	uid_string = g_conf->setuser;
+  else {
+    // drop privileges?
+
+    if (g_conf->setgroup.length() ||
+        g_conf->setuser.length()) {
```
Comment (liewegas) [inline src/global/global_init.cc]: i think if you switch this to else if () { .. you won't have to reindent this whole block?

**Comment by Liuchang0812** [comment]
hi, @liewegas

I will have a try. thanks

2017-03-22 20:21 GMT+08:00 Sage Weil <notifications@github.com>:

**Comment by liewegas** [comment]
not sure about 3; but an easier step would be to rename the member varaible for SAFE_OPTION to __safe_##name## or something so that someone trying to do g_conf->some_safe_option will fail to compile.

Anyway, this series looks good to me!

**Pull request merged**